Primero cargaré todos los datos para revisar que todas las columnas tengan los datos correctos tanto en formato como en información y después realizaré la unión de todos los años para tratarlo en conjunto.

In [2]:
#Importo las librerías a usar
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)

In [6]:
#Importo 2013 usando encoding latin-1 para que los datos aparezcan en español
datos_2013 = pd.read_csv("datos-2013.csv", sep = ";", encoding = "latin-1")
datos_2013.head(5)



,CÓDIGO ÓRGÃO SUPERIOR,NOME ÓRGÃO SUPERIOR,CÓDIGO ÓRGÃO,NOME ÓRGÃO,CÓDIGO UNIDADE GESTORA,NOME UNIDADE GESTORA,CATEGORIA ECONÔMICA,ORIGEM RECEITA,ESPÉCIE RECEITA,DETALHAMENTO,VALOR PREVISTO ATUALIZADO,VALOR LANÇADO,VALOR REALIZADO,PERCENTUAL REALIZADO,DATA LANÇAMENTO,ANO EXERCÍCIO
0,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Bens, Direitos e Valores Incorporados ao Patr",REC.DIVIDA ATIVA NAO TRIBUTARIA DE OUTRAS REC,"0,00","0,00","1297,13","0,00",31/12/2013,2013
1,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Indenizações, restituições e ressarcimentos",RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES,"0,00","0,00","26666621,42","0,00",31/12/2013,2013
2,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Multas administrativas, contratuais e judicia",OUTRAS MULTAS E JUROS DE MORA,"0,00","0,00","301251,13","0,00",31/12/2013,2013
3,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Bens, Direitos e Valores Incorporados ao Patr",REC.DIV.ATIVA POR INFRAÇÃO ADMINISTRATIVA,"0,00","0,00","1855,58","0,00",31/12/2013,2013
4,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Indenizações, restituições e ressarcimentos",OUTRAS RESTITUICOES,"0,00","0,00","52140,68","0,00",31/12/2013,2013


In [15]:
#Revisare los formatos e información básica de las columnas
datos_2013.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4498 entries, 0 to 4497
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      4498 non-null   int64         
 1   NOME ÓRGÃO SUPERIOR        4498 non-null   object        
 2   CÓDIGO ÓRGÃO               4498 non-null   int64         
 3   NOME ÓRGÃO                 4498 non-null   object        
 4   CÓDIGO UNIDADE GESTORA     4498 non-null   int64         
 5   NOME UNIDADE GESTORA       4498 non-null   object        
 6   CATEGORIA ECONÔMICA        4498 non-null   object        
 7   ORIGEM RECEITA             4498 non-null   object        
 8   ESPÉCIE RECEITA            4498 non-null   object        
 9   DETALHAMENTO               4498 non-null   object        
 10  VALOR PREVISTO ATUALIZADO  4498 non-null   object        
 11  VALOR LANÇADO              4498 non-null   object        
 12  VALOR 

In [ ]:
#tenemos varias columnas que no tienen el formato ideal (fechas como objeto, números como objeto...) Vamos a modificarlo
datos_2013["DATA LANÇAMENTO"] = pd.to_datetime(datos_2013["DATA LANÇAMENTO"], format="%d/%m/%Y")
datos_2013.head(2)
print(datos_2013["DATA LANÇAMENTO"].dtypes)

datetime64[ns]


In [24]:
#Ahora cambio las columnas con texto a número que me interesan
datos_2013["VALOR PREVISTO ATUALIZADO"]= datos_2013["VALOR PREVISTO ATUALIZADO"].astype(float)

In [26]:
datos_2013["VALOR LANÇADO"]= datos_2013["VALOR LANÇADO"].str.replace(",",".")
datos_2013["VALOR LANÇADO"]= datos_2013["VALOR LANÇADO"].astype(float)

In [27]:
datos_2013["VALOR REALIZADO"]= datos_2013["VALOR REALIZADO"].str.replace(",",".")
datos_2013["VALOR REALIZADO"]= datos_2013["VALOR REALIZADO"].astype(float)

In [28]:
datos_2013["PERCENTUAL REALIZADO"]= datos_2013["PERCENTUAL REALIZADO"].str.replace(",",".")
datos_2013["PERCENTUAL REALIZADO"]= datos_2013["PERCENTUAL REALIZADO"].astype(float)

In [29]:
#vuelvo a ver como quedan las columnas
datos_2013.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4498 entries, 0 to 4497
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      4498 non-null   int64         
 1   NOME ÓRGÃO SUPERIOR        4498 non-null   object        
 2   CÓDIGO ÓRGÃO               4498 non-null   int64         
 3   NOME ÓRGÃO                 4498 non-null   object        
 4   CÓDIGO UNIDADE GESTORA     4498 non-null   int64         
 5   NOME UNIDADE GESTORA       4498 non-null   object        
 6   CATEGORIA ECONÔMICA        4498 non-null   object        
 7   ORIGEM RECEITA             4498 non-null   object        
 8   ESPÉCIE RECEITA            4498 non-null   object        
 9   DETALHAMENTO               4498 non-null   object        
 10  VALOR PREVISTO ATUALIZADO  4498 non-null   float64       
 11  VALOR LANÇADO              4498 non-null   float64       
 12  VALOR 

In [ ]:
#reviso si existen duplicados y parece que no existen filas duplicadas
duplicados_df_2013 = datos_2013.duplicated().sum()
duplicados_df_2013


np.int64(0)

Ya tenemos limpio 2013, no hay valores ni nulos ni duplicados y todas las columnas tienen 4498 registros

Hago lo mismo para el resto de Data Frames y como son todos iguales, voy a crear una función para solo tener que lanzarla en resto de data frames



In [45]:
#Creo la función para limpiar los datos
def limpieza_datos(df):
    #crearé una copia para no cargarme el original
    df_limpio = df.copy()
    #corrige la fecha 
    df_limpio["DATA LANÇAMENTO"] = pd.to_datetime(df_limpio["DATA LANÇAMENTO"], format="%d/%m/%Y")
    
    # Convertir columnas monetarias y porcentuales
    columnas_a_convertir = ["VALOR PREVISTO ATUALIZADO", "PERCENTUAL REALIZADO", 
                            "VALOR LANÇADO", "VALOR REALIZADO"]
    
    for columna in columnas_a_convertir:
        if columna in df_limpio.columns:
            # Reemplazar comas por puntos
            df_limpio[columna] = df_limpio[columna].astype(str).str.replace(",", ".")
            # Convertir a float
            df_limpio[columna] = df_limpio[columna].astype(float)
    
    return df_limpio

In [49]:
#También creo una función para comprobar duplicados
def comprueba_dup(df):
    duplicados_df = df.duplicated().sum()
    return  print(duplicados_df)

In [46]:
#Importo 2014 y le aplico la función 
datos_2014 = pd.read_csv("datos-2014.csv", sep = ";", encoding = "latin-1")
datos_2014.head(3)


,CÓDIGO ÓRGÃO SUPERIOR,NOME ÓRGÃO SUPERIOR,CÓDIGO ÓRGÃO,NOME ÓRGÃO,CÓDIGO UNIDADE GESTORA,NOME UNIDADE GESTORA,CATEGORIA ECONÔMICA,ORIGEM RECEITA,ESPÉCIE RECEITA,DETALHAMENTO,VALOR PREVISTO ATUALIZADO,VALOR LANÇADO,VALOR REALIZADO,PERCENTUAL REALIZADO,DATA LANÇAMENTO,ANO EXERCÍCIO
0,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,Demais receitas correntes,RECEITA DE HONORÁRIOS DE ADVOGADOS,"0,00","0,00","46699948,08","0,00",31/12/2014,2014
1,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Multas administrativas, contratuais e judicia",MULTAS E JUROS PREVISTOS EM CONTRATOS,"0,00","0,00","15920,83","0,00",31/12/2014,2014
2,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Multas administrativas, contratuais e judicia",OUTRAS MULTAS E JUROS DE MORA,"0,00","0,00","2195899,92","0,00",31/12/2014,2014


In [47]:
datos_2014 = limpieza_datos(datos_2014)

In [48]:
datos_2014.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4553 entries, 0 to 4552
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      4553 non-null   int64         
 1   NOME ÓRGÃO SUPERIOR        4553 non-null   object        
 2   CÓDIGO ÓRGÃO               4553 non-null   int64         
 3   NOME ÓRGÃO                 4553 non-null   object        
 4   CÓDIGO UNIDADE GESTORA     4553 non-null   int64         
 5   NOME UNIDADE GESTORA       4553 non-null   object        
 6   CATEGORIA ECONÔMICA        4553 non-null   object        
 7   ORIGEM RECEITA             4553 non-null   object        
 8   ESPÉCIE RECEITA            4553 non-null   object        
 9   DETALHAMENTO               4553 non-null   object        
 10  VALOR PREVISTO ATUALIZADO  4553 non-null   float64       
 11  VALOR LANÇADO              4553 non-null   float64       
 12  VALOR 

In [50]:
comprueba_dup(datos_2014)

0


Datos 2014 ya estaría limpio y sin duplicados, continuo haciendo lo mismo para el resto de datos

In [51]:
datos_2015 = pd.read_csv("datos-2015.csv", sep = ";", encoding = "latin-1")
datos_2015 = limpieza_datos(datos_2015)
datos_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4523 entries, 0 to 4522
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      4523 non-null   int64         
 1   NOME ÓRGÃO SUPERIOR        4523 non-null   object        
 2   CÓDIGO ÓRGÃO               4523 non-null   int64         
 3   NOME ÓRGÃO                 4523 non-null   object        
 4   CÓDIGO UNIDADE GESTORA     4523 non-null   int64         
 5   NOME UNIDADE GESTORA       4523 non-null   object        
 6   CATEGORIA ECONÔMICA        4523 non-null   object        
 7   ORIGEM RECEITA             4523 non-null   object        
 8   ESPÉCIE RECEITA            4523 non-null   object        
 9   DETALHAMENTO               4523 non-null   object        
 10  VALOR PREVISTO ATUALIZADO  4523 non-null   float64       
 11  VALOR LANÇADO              4523 non-null   float64       
 12  VALOR 

In [52]:
comprueba_dup(datos_2015)

0


In [53]:
datos_2016 = pd.read_csv("datos-2016.csv", sep = ";", encoding = "latin-1")
datos_2016 = limpieza_datos(datos_2016)
datos_2016.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194533 entries, 0 to 194532
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      194533 non-null  int64         
 1   NOME ÓRGÃO SUPERIOR        194533 non-null  object        
 2   CÓDIGO ÓRGÃO               194533 non-null  int64         
 3   NOME ÓRGÃO                 194533 non-null  object        
 4   CÓDIGO UNIDADE GESTORA     194533 non-null  int64         
 5   NOME UNIDADE GESTORA       194533 non-null  object        
 6   CATEGORIA ECONÔMICA        194533 non-null  object        
 7   ORIGEM RECEITA             194533 non-null  object        
 8   ESPÉCIE RECEITA            194533 non-null  object        
 9   DETALHAMENTO               194533 non-null  object        
 10  VALOR PREVISTO ATUALIZADO  194533 non-null  float64       
 11  VALOR LANÇADO              194533 non-null  float64 

In [54]:
comprueba_dup(datos_2016)

0


In [55]:
datos_2017 = pd.read_csv("datos-2017.csv", sep = ";", encoding = "latin-1")
datos_2017 = limpieza_datos(datos_2017)
datos_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190479 entries, 0 to 190478
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      190479 non-null  int64         
 1   NOME ÓRGÃO SUPERIOR        190479 non-null  object        
 2   CÓDIGO ÓRGÃO               190479 non-null  int64         
 3   NOME ÓRGÃO                 190479 non-null  object        
 4   CÓDIGO UNIDADE GESTORA     190479 non-null  int64         
 5   NOME UNIDADE GESTORA       190479 non-null  object        
 6   CATEGORIA ECONÔMICA        190479 non-null  object        
 7   ORIGEM RECEITA             190479 non-null  object        
 8   ESPÉCIE RECEITA            190479 non-null  object        
 9   DETALHAMENTO               190479 non-null  object        
 10  VALOR PREVISTO ATUALIZADO  190479 non-null  float64       
 11  VALOR LANÇADO              190479 non-null  float64 

In [56]:
comprueba_dup(datos_2017)

0


In [57]:
datos_2018 = pd.read_csv("datos-2018.csv", sep = ";", encoding = "latin-1")
datos_2018 = limpieza_datos(datos_2018)
datos_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173944 entries, 0 to 173943
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      173944 non-null  int64         
 1   NOME ÓRGÃO SUPERIOR        173944 non-null  object        
 2   CÓDIGO ÓRGÃO               173944 non-null  int64         
 3   NOME ÓRGÃO                 173944 non-null  object        
 4   CÓDIGO UNIDADE GESTORA     173944 non-null  int64         
 5   NOME UNIDADE GESTORA       173944 non-null  object        
 6   CATEGORIA ECONÔMICA        173944 non-null  object        
 7   ORIGEM RECEITA             173944 non-null  object        
 8   ESPÉCIE RECEITA            173944 non-null  object        
 9   DETALHAMENTO               173944 non-null  object        
 10  VALOR PREVISTO ATUALIZADO  173944 non-null  float64       
 11  VALOR LANÇADO              173944 non-null  float64 

In [58]:
comprueba_dup(datos_2018)

0


In [59]:
datos_2019 = pd.read_csv("datos-2019.csv", sep = ";", encoding = "latin-1")
datos_2019 = limpieza_datos(datos_2019)
datos_2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176828 entries, 0 to 176827
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      176828 non-null  int64         
 1   NOME ÓRGÃO SUPERIOR        176828 non-null  object        
 2   CÓDIGO ÓRGÃO               176828 non-null  int64         
 3   NOME ÓRGÃO                 176828 non-null  object        
 4   CÓDIGO UNIDADE GESTORA     176828 non-null  int64         
 5   NOME UNIDADE GESTORA       176828 non-null  object        
 6   CATEGORIA ECONÔMICA        176828 non-null  object        
 7   ORIGEM RECEITA             176828 non-null  object        
 8   ESPÉCIE RECEITA            176828 non-null  object        
 9   DETALHAMENTO               176828 non-null  object        
 10  VALOR PREVISTO ATUALIZADO  176828 non-null  float64       
 11  VALOR LANÇADO              176828 non-null  float64 

In [61]:
comprueba_dup(datos_2019)

0


In [62]:
datos_2020 = pd.read_csv("datos-2020.csv", sep = ";", encoding = "latin-1")
datos_2020 = limpieza_datos(datos_2020)
datos_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 142348 entries, 0 to 142347
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      142348 non-null  int64         
 1   NOME ÓRGÃO SUPERIOR        142348 non-null  object        
 2   CÓDIGO ÓRGÃO               142348 non-null  int64         
 3   NOME ÓRGÃO                 142348 non-null  object        
 4   CÓDIGO UNIDADE GESTORA     142348 non-null  int64         
 5   NOME UNIDADE GESTORA       142348 non-null  object        
 6   CATEGORIA ECONÔMICA        142348 non-null  object        
 7   ORIGEM RECEITA             142348 non-null  object        
 8   ESPÉCIE RECEITA            142348 non-null  object        
 9   DETALHAMENTO               142348 non-null  object        
 10  VALOR PREVISTO ATUALIZADO  142348 non-null  float64       
 11  VALOR LANÇADO              142348 non-null  float64 

In [63]:
comprueba_dup(datos_2020)

0


In [64]:
datos_2021 = pd.read_csv("datos-2021.csv", sep = ";", encoding = "latin-1")
datos_2021 = limpieza_datos(datos_2021)
datos_2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134593 entries, 0 to 134592
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      134593 non-null  int64         
 1   NOME ÓRGÃO SUPERIOR        134593 non-null  object        
 2   CÓDIGO ÓRGÃO               134593 non-null  int64         
 3   NOME ÓRGÃO                 134593 non-null  object        
 4   CÓDIGO UNIDADE GESTORA     134593 non-null  int64         
 5   NOME UNIDADE GESTORA       134593 non-null  object        
 6   CATEGORIA ECONÔMICA        134593 non-null  object        
 7   ORIGEM RECEITA             134593 non-null  object        
 8   ESPÉCIE RECEITA            134593 non-null  object        
 9   DETALHAMENTO               134593 non-null  object        
 10  VALOR PREVISTO ATUALIZADO  134593 non-null  float64       
 11  VALOR LANÇADO              134593 non-null  float64 

In [66]:
comprueba_dup(datos_2021)

0


Hemos visto que en estos años existen valores nulos dentro de la columna "Data Lancamento". Trataremos estos valores en conjunto con el data frame definitivo. Uno todos los data frames anuales en uno conjunto.

In [67]:
datos_brasil = pd.concat([datos_2013, datos_2014, datos_2015, datos_2016,datos_2017, datos_2018, datos_2019, datos_2020, datos_2021], axis = 0, ignore_index = True)
datos_brasil.head(2)

,CÓDIGO ÓRGÃO SUPERIOR,NOME ÓRGÃO SUPERIOR,CÓDIGO ÓRGÃO,NOME ÓRGÃO,CÓDIGO UNIDADE GESTORA,NOME UNIDADE GESTORA,CATEGORIA ECONÔMICA,ORIGEM RECEITA,ESPÉCIE RECEITA,DETALHAMENTO,VALOR PREVISTO ATUALIZADO,VALOR LANÇADO,VALOR REALIZADO,PERCENTUAL REALIZADO,DATA LANÇAMENTO,ANO EXERCÍCIO
0,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Bens, Direitos e Valores Incorporados ao Patr",REC.DIVIDA ATIVA NAO TRIBUTARIA DE OUTRAS REC,0.0,0.0,1297.13,0.0,2013-12-31,2013
1,63000,Advocacia-Geral da União,63000,Advocacia-Geral da União - Unidades com víncul...,110060,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Indenizações, restituições e ressarcimentos",RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES,0.0,0.0,26666621.42,0.0,2013-12-31,2013


In [68]:
datos_brasil.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      1026299 non-null  int64         
 1   NOME ÓRGÃO SUPERIOR        1026299 non-null  object        
 2   CÓDIGO ÓRGÃO               1026299 non-null  int64         
 3   NOME ÓRGÃO                 1026299 non-null  object        
 4   CÓDIGO UNIDADE GESTORA     1026299 non-null  int64         
 5   NOME UNIDADE GESTORA       1026299 non-null  object        
 6   CATEGORIA ECONÔMICA        1026299 non-null  object        
 7   ORIGEM RECEITA             1026299 non-null  object        
 8   ESPÉCIE RECEITA            1026299 non-null  object        
 9   DETALHAMENTO               1026299 non-null  object        
 10  VALOR PREVISTO ATUALIZADO  1026299 non-null  float64       
 11  VALOR LANÇADO              1026299 no

Una vez que tenemos todo unido, vamos a realizar un pequeño análisis EDA para ver que nos dicen los datos

In [73]:
datos_brasil.describe().T

,count,mean,min,25%,50%,75%,max,std
CÓDIGO ÓRGÃO SUPERIOR,1026299.0,32050.523288,20000.0,25000.0,26000.0,39000.0,81000.0,10612.547085
CÓDIGO ÓRGÃO,1026299.0,31435.161341,20101.0,25201.0,26298.0,36000.0,91214.0,10382.647574
CÓDIGO UNIDADE GESTORA,1026299.0,231293.743317,110005.0,154003.0,170013.0,250088.0,913001.0,144178.420681
VALOR PREVISTO ATUALIZADO,1026299.0,27611421.049683,-214773743.0,0.0,0.0,0.0,1603521711208.0,3393367175.127452
VALOR LANÇADO,1026299.0,5242806.106963,-148347267246.910004,0.0,0.0,0.0,357160677863.969971,810883695.355811
VALOR REALIZADO,1026299.0,24980422.309138,-156285934188.029999,260.905,3022.32,40853.855,771117711060.949951,1406892548.8405
PERCENTUAL REALIZADO,1026299.0,80.313812,-114552.0,0.0,0.0,0.0,72363772.0,71481.112902
DATA LANÇAMENTO,1025721,2018-09-20 12:23:18.569786368,2013-12-31 00:00:00,2017-03-31 00:00:00,2018-08-27 00:00:00,2020-02-12 00:00:00,2021-12-03 00:00:00,NaN
ANO EXERCÍCIO,1026299.0,2018.225555,2013.0,2017.0,2018.0,2020.0,2021.0,1.737966


Vemos que existen valores nulos en la fecha de lanzamiento, haremos un pequeño análisis para ver cuales son y si los podemos rellenar de alguna forma

In [78]:
filtro_nulos = datos_brasil[datos_brasil["DATA LANÇAMENTO"].isnull()]
filtro_nulos.head(5)

,CÓDIGO ÓRGÃO SUPERIOR,NOME ÓRGÃO SUPERIOR,CÓDIGO ÓRGÃO,NOME ÓRGÃO,CÓDIGO UNIDADE GESTORA,NOME UNIDADE GESTORA,CATEGORIA ECONÔMICA,ORIGEM RECEITA,ESPÉCIE RECEITA,DETALHAMENTO,VALOR PREVISTO ATUALIZADO,VALOR LANÇADO,VALOR REALIZADO,PERCENTUAL REALIZADO,DATA LANÇAMENTO,ANO EXERCÍCIO
31715,24000,"Ministério da Ciência, Tecnologia, Inovações",20501,Conselho Nacional de Desenvolvimento Científic...,364102,CONSELHO NAC DE DESENV CIENT E TECNOLOGICO,Receitas de Capital,Operações de Crédito,Operações de crédito - mercado interno,EMPRÉSTIMOS COMPULSÓRIOS,0.0,3.665722e+04,0.0,0.0,NaT,2016
50971,25000,Ministério da Economia,25000,Ministério da Economia - Unidades com vínculo ...,170010,SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB,Receitas Correntes,"Impostos, Taxas e Contribuições de Melhoria",Impostos,IOF - COMERCIALIZACAO DO OURO,0.0,2.490230e+05,0.0,0.0,NaT,2016
51032,25000,Ministério da Economia,25000,Ministério da Economia - Unidades com vínculo ...,170010,SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB,Receitas Correntes,"Impostos, Taxas e Contribuições de Melhoria",Impostos,RECEITA DE PARCELAMENTOS - IMPOSTO EXPORTACAO,0.0,3.287169e+07,0.0,0.0,NaT,2016
51747,25000,Ministério da Economia,25000,Ministério da Economia - Unidades com vínculo ...,170010,SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB,Receitas Correntes,"Impostos, Taxas e Contribuições de Melhoria",Impostos,RETIDO NAS FONTES - CAPITAL,0.0,1.753181e+09,0.0,0.0,NaT,2016
51977,25000,Ministério da Economia,25000,Ministério da Economia - Unidades com vínculo ...,170010,SECRETARIA DA RECEITA FEDERAL DO BRASIL-RFB,Receitas Correntes,"Impostos, Taxas e Contribuições de Melhoria",Impostos,IRRF-TRABALHO,0.0,2.242782e+08,0.0,0.0,NaT,2016


In [ ]:
filtro_nulos_año = filtro_nulos.groupby(filtro_nulos["ANO EXERCÍCIO"]).count("")
filtro_nulos_año

,CÓDIGO ÓRGÃO SUPERIOR,NOME ÓRGÃO SUPERIOR,CÓDIGO ÓRGÃO,NOME ÓRGÃO,CÓDIGO UNIDADE GESTORA,NOME UNIDADE GESTORA,CATEGORIA ECONÔMICA,ORIGEM RECEITA,ESPÉCIE RECEITA,DETALHAMENTO,VALOR PREVISTO ATUALIZADO,VALOR LANÇADO,VALOR REALIZADO,PERCENTUAL REALIZADO,DATA LANÇAMENTO
ANO EXERCÍCIO,,,,,,,,,,,,,,,
2016,82,82,82,82,82,82,82,82,82,82,82,82,82,82,0
2017,86,86,86,86,86,86,86,86,86,86,86,86,86,86,0
2018,115,115,115,115,115,115,115,115,115,115,115,115,115,115,0
2019,86,86,86,86,86,86,86,86,86,86,86,86,86,86,0
2020,86,86,86,86,86,86,86,86,86,86,86,86,86,86,0
2021,123,123,123,123,123,123,123,123,123,123,123,123,123,123,0


In [ ]:
#Vemos que empiezan a partir de 2016, contaré el valor de los nulos sobre el total
medida = datos_brasil.shape
porcentaje_nulos = datos_brasil["DATA LANÇAMENTO"].isnull().sum() / medida[0] * 100
porcentaje_nulos = float(porcentaje_nulos)
porcentaje_nulos

0.05631887003689958

el porcentaje de nulos es del 0.05% sobre el total de registro y está repartido en varios años por lo que decido quitar los valores nulos ya que no podemos obtenerlos de otras columnas y no son significativos para nuestro análisis ni parece que vengan de un solo año.

In [110]:
brasil_datos_limpio = datos_brasil.dropna(subset=["DATA LANÇAMENTO"])
brasil_datos_limpio.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1025721 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   CÓDIGO ÓRGÃO SUPERIOR      1025721 non-null  int64         
 1   NOME ÓRGÃO SUPERIOR        1025721 non-null  object        
 2   CÓDIGO ÓRGÃO               1025721 non-null  int64         
 3   NOME ÓRGÃO                 1025721 non-null  object        
 4   CÓDIGO UNIDADE GESTORA     1025721 non-null  int64         
 5   NOME UNIDADE GESTORA       1025721 non-null  object        
 6   CATEGORIA ECONÔMICA        1025721 non-null  object        
 7   ORIGEM RECEITA             1025721 non-null  object        
 8   ESPÉCIE RECEITA            1025721 non-null  object        
 9   DETALHAMENTO               1025721 non-null  object        
 10  VALOR PREVISTO ATUALIZADO  1025721 non-null  float64       
 11  VALOR LANÇADO              1025721 non-nul

In [111]:
#Ahora que no hay nulos, vamos a ver la naturaleza de las columnas y si son coherentes
brasil_datos_limpio.describe().T


,count,mean,min,25%,50%,75%,max,std
CÓDIGO ÓRGÃO SUPERIOR,1025721.0,32053.627643,20000.0,25000.0,26000.0,39000.0,81000.0,10614.192759
CÓDIGO ÓRGÃO,1025721.0,31438.190265,20101.0,25201.0,26298.0,36000.0,91214.0,10384.494245
CÓDIGO UNIDADE GESTORA,1025721.0,231320.807146,110005.0,154003.0,170013.0,250088.0,913001.0,144211.265977
VALOR PREVISTO ATUALIZADO,1025721.0,27626980.252787,-214773743.0,0.0,0.0,0.0,1603521711208.0,3394323069.571367
VALOR LANÇADO,1025721.0,2712342.560349,-148347267246.910004,0.0,0.0,0.0,357160677863.969971,698431477.235606
VALOR REALIZADO,1025721.0,24994498.928506,-156285934188.029999,262.5,3030.0,40932.5,771117711060.949951,1407288764.632859
PERCENTUAL REALIZADO,1025721.0,80.359069,-114552.0,0.0,0.0,0.0,72363772.0,71501.25008
DATA LANÇAMENTO,1025721,2018-09-20 12:23:18.569787136,2013-12-31 00:00:00,2017-03-31 00:00:00,2018-08-27 00:00:00,2020-02-12 00:00:00,2021-12-03 00:00:00,NaN
ANO EXERCÍCIO,1025721.0,2018.225315,2013.0,2017.0,2018.0,2020.0,2021.0,1.737948


In [115]:
#Convertiré las columnas de codigo a objeto ya que tiene más sentido puesto que los códigos van relacionados con los nombres
brasil_datos_limpio["CÓDIGO ÓRGÃO SUPERIOR"] = brasil_datos_limpio["CÓDIGO ÓRGÃO SUPERIOR"].astype("object")
brasil_datos_limpio["CÓDIGO ÓRGÃO"] = brasil_datos_limpio["CÓDIGO ÓRGÃO"].astype("object")
brasil_datos_limpio["CÓDIGO UNIDADE GESTORA"] = brasil_datos_limpio["CÓDIGO UNIDADE GESTORA"].astype("object")

C:\Users\ganad\AppData\Local\Temp\ipykernel_16668\1481096915.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  brasil_datos_limpio["CÓDIGO ÓRGÃO SUPERIOR"] = brasil_datos_limpio["CÓDIGO ÓRGÃO SUPERIOR"].astype("object")
C:\Users\ganad\AppData\Local\Temp\ipykernel_16668\1481096915.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  brasil_datos_limpio["CÓDIGO ÓRGÃO"] = brasil_datos_limpio["CÓDIGO ÓRGÃO"].astype("object")
C:\Users\ganad\AppData\Local\Temp\ipykernel_16668\1481096915.py:4: SettingWithCopyWarn

In [116]:
brasil_datos_limpio.describe(include=object).T

,count,unique,top,freq
CÓDIGO ÓRGÃO SUPERIOR,1025721,25,26000,361867
NOME ÓRGÃO SUPERIOR,1025721,25,Ministério da Educação,361867
CÓDIGO ÓRGÃO,1025721,291,25000,119345
NOME ÓRGÃO,1025721,287,Ministério da Economia - Unidades com vínculo ...,119345
CÓDIGO UNIDADE GESTORA,1025721,364,170013,108481
NOME UNIDADE GESTORA,1025721,356,SETORIAL ORCAMENTARIA E FINANCEIRA / ME,108481
CATEGORIA ECONÔMICA,1025721,5,Receitas Correntes,979101
ORIGEM RECEITA,1025721,15,Outras Receitas Correntes,322884
ESPÉCIE RECEITA,1025721,63,Serviços Administrativos e Comerciais Gerais,278157
DETALHAMENTO,1025721,1886,SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.,159039


Las columnas de codigo orgao y nombre de orgao deberían tener los mismos valores vemos que hay algunas organizaciones con varios códigos, lo mismo pasa con la unidad gestora. Como lo que nos importa son los nombres, eliminare las columnas de códigos para evitar confusiónes 

In [117]:
brasil_datos_limpio = brasil_datos_limpio.drop(columns=['CÓDIGO ÓRGÃO SUPERIOR', 'CÓDIGO ÓRGÃO', 'CÓDIGO UNIDADE GESTORA'])
brasil_datos_limpio.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1025721 entries, 0 to 1026298
Data columns (total 13 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   NOME ÓRGÃO SUPERIOR        1025721 non-null  object        
 1   NOME ÓRGÃO                 1025721 non-null  object        
 2   NOME UNIDADE GESTORA       1025721 non-null  object        
 3   CATEGORIA ECONÔMICA        1025721 non-null  object        
 4   ORIGEM RECEITA             1025721 non-null  object        
 5   ESPÉCIE RECEITA            1025721 non-null  object        
 6   DETALHAMENTO               1025721 non-null  object        
 7   VALOR PREVISTO ATUALIZADO  1025721 non-null  float64       
 8   VALOR LANÇADO              1025721 non-null  float64       
 9   VALOR REALIZADO            1025721 non-null  float64       
 10  PERCENTUAL REALIZADO       1025721 non-null  float64       
 11  DATA LANÇAMENTO            1025721 non-nul

In [118]:
#Guardo el Data frame final para usarlo después
brasil_datos_limpio.to_csv("brasil_datos_limpio",index=False)

In [121]:
datos_OK = pd.read_csv("brasil_datos_limpio", sep = ",", encoding = "latin-1")
datos_OK.head(3)

,NOME ÓRGÃO SUPERIOR,NOME ÓRGÃO,NOME UNIDADE GESTORA,CATEGORIA ECONÔMICA,ORIGEM RECEITA,ESPÉCIE RECEITA,DETALHAMENTO,VALOR PREVISTO ATUALIZADO,VALOR LANÇADO,VALOR REALIZADO,PERCENTUAL REALIZADO,DATA LANÇAMENTO,ANO EXERCÍCIO
0,Advocacia-Geral da União,Advocacia-Geral da União - Unidades com vínc...,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Bens, Direitos e Valores Incorporados ao Patr",REC.DIVIDA ATIVA NAO TRIBUTARIA DE OUTRAS REC,0.0,0.0,1297.13,0.0,2013-12-31,2013
1,Advocacia-Geral da União,Advocacia-Geral da União - Unidades com vínc...,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Indenizações, restituições e ressarcimentos",RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES,0.0,0.0,26666621.42,0.0,2013-12-31,2013
2,Advocacia-Geral da União,Advocacia-Geral da União - Unidades com vínc...,COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU,Receitas Correntes,Outras Receitas Correntes,"Multas administrativas, contratuais e judicia",OUTRAS MULTAS E JUROS DE MORA,0.0,0.0,301251.13,0.0,2013-12-31,2013
